metal/end_model/loss.py

```diff
@@ -18,17 +18,24 @@ class SoftCrossEntropyLoss(nn.Module):
         target: An [n, k] float tensor of target probabilities
     """
 
-    def __init__(self, weight=None, reduction="elementwise_mean"):
+    def __init__(
+        self, weight=None, reduction="elementwise_mean", use_cuda=False
+    ):
         super().__init__()
         assert weight is None or isinstance(weight, torch.FloatTensor)
         self.weight = weight
         self.reduction = reduction
+        self.use_cuda = use_cuda
```

Comment:
@ajratner my remembrance is that it didn't work because of these two lines:

`cum_losses = torch.zeros(n)`
`cls_idx = torch.full((n,), y, dtype=torch.long)`

each of which define a new variable inside the module, which by default was on the CPU.  I manually switched that to GPU to make sure that did not cause a problem.  We also put the rest of the module on the GPU as well when we do `criteria.cuda()` in `get_loss_fn`.  Make sense?

Reply:
Got it!

Reply:
I think there's a better way. torch.zeros() may default to cpu, but the torch.Tensor.new_zeros() will default to the same device as whatever tensor you call it on (same with .new_full()). So let's just have both cum_losses and cls_idx be on the same device as input using input.new_zeros() and input.new_full().

Reply:
Good call!